trieste/models/interfaces.py

```diff
@@ -272,25 +272,104 @@ def reparam_sampler(self, num_samples: int) -> ReparametrizationSampler:
         as their reparameterization sampler.
 
         :param num_samples: The desired number of samples.
-        :return: The reparameterization sampler.
+        :return: The reparametrization sampler.
         :raise NotImplementedError: If the models in the stack do not share the
             same :meth:`reparam_sampler`.
         """
 
-        samplers = [model.reparam_sampler(num_samples) for model in self._models]
-
-        if all(isinstance(sampler, BatchReparametrizationSampler) for sampler in samplers):
-            return BatchReparametrizationSampler(
-                num_samples, self
-            )  # return the shared sampler rebuilt for the whole model stack
-        else:
+        sampler_types = [type(model.reparam_sampler(1)) for model in self._models]
+        unique_sampler_types = set(sampler_types)
+        if len(unique_sampler_types) > 1:
```

Comment:
I'm confused. Did you make the change and then revert it?

(also as an aside the code should actually check for type equality, not instance, as the optimization doesn't support using subtypes of BatchReparameterizationSampler)

```python
if all(type(sampler) is BatchReparameterizationSampler for sampler in samplers):
```